janus_server/src/aggregator.rs

```diff
@@ -1091,11 +1086,68 @@ impl VdafOps {
             )
             .await?;
 
+        // At the moment we handle the AggregateShareReq, there could be some incomplete aggregation
+        // jobs whose results not been accumulated into the batch unit aggregations we just queried
+        // from the datastore, meaning we will aggregate over an incomplete view of data, which:
+        //
+        //  * reduces fidelity of the resulting aggregates,
+        //  * could cause us to fail to meet the minimum batch size for the task,
+        //  * or for particularly pathological timing, could cause us to aggregate a different set
+        //    of reports than the leader did (though the checksum will detect this).
+        //
+        // There's not much the helper can do about this, because an aggregate job might be
+        // unfinished because it's waiting on an aggregate sub-protocol message that is never coming
+        // because the leader has abandoned that job.
+        //
+        // Thus the helper has no choice but to assume that any unfinished aggregation jobs were
+        // intentionally abandoned by the leader and service the aggregate share request with
+        // whatever batch unit aggregations are available now.
+        //
+        // See issue #104 for more discussion.
+
         let mut total_report_count = 0;
         let mut total_checksum = [0u8; 32];
         let mut total_aggregate_share: Option<A::AggregateShare> = None;
 
         for batch_unit_aggregation in &batch_unit_aggregations {
+            // Check how many rows in aggregate_share_jobs have a batch interval that includes this
+            // batch unit. Each such row consumes one unit of batch lifetime (§4.6).
+            //
+            // We have to check each batch unit interval separately rather than checking once how
+            // many times aggregate_share_req.batch_interval overlaps with any row. Suppose we had:
+            //
+            //   * task.max_batch_lifetime = 2,
+            //   * an AggregateShareReq.batch interval that spans two batch units,
+            //   * and that each of those batch units has been collected once before.
+            //
+            // A further AggregateShareReq including either or both of the units is permissible, but
+            // if we queried how many rows overlap with that interval, we would get 2 and refuse the
+            // request. We must check the unit intervals individually to notice that each has enough
+            // remaining lifetime to permit the share request.
+            let unit_interval = Interval::new(
+                batch_unit_aggregation.unit_interval_start,
+                task.min_batch_duration,
+            )?;
+            let consumed_batch_lifetime = tx
```

Comment:
These checks are independent, and can be done with concurrency: we could start each of the futures and join on them using e.g. `try_join_all`, then run the checks on each result.

An alternate strategy would be to update the datastore operation to return all of the necessary counts by batch unit interval rather than being a one-off, e.g. something like `SELECT batch_unit_start, COUNT(*) AS count FROM ... GROUP BY batch_unit_start`. I'm not sure which would be easier to implement.

Reply:
I'm going to try to write this as a single SQL query using [`WHERE ANY`](https://www.postgresql.org/docs/current/functions-comparisons.html#id-1.5.8.30.16).